### DISCUSSION TITLE
Transpiler!

### DISCUSSION THREAD

**Discussion by jwoehr** [Qiskit/.github#3] [Ideas]
The transpiler works well enough, but where is it going? One of the most hidden and most interesting aspects of Qiskit.

**Comment by 1ucian0** [comment]
One aspect I think it can be improved is a more coupling-map/noise awareness across the passes. For example, when decomposing very abstract boxes such as VQE, knowing the coupling map might reduce the amount of swaps you need down the line.

**Reply by jwoehr** [reply]
It's an intrinsically difficult problem where the only proof of optimal solution would be to evaluate an exhaustive number of possibilities. Pretty much a traveling salesman kind of problem. The behavior of the transpiler shows that it is subject to getting trapped in local minima.

**Comment by jwoehr** [comment]
@1ucian0 (Dr. Bello) and I chatted today about a pet idea of mine:

### As much of the work of the transpiler as can run on the quantum device _should_ run on the quantum device.

Using quantum computation to route and optimize the quantum circuit could have an echo effect of both improving the transpiler and helping us perfect quantum routing and optimization algorithms.

What Qiskit calls "the transpiler" is a bundle of components performing various tasks, most of which currently are not amenable to porting to the quantum device itself. Currently the quantum device cannot, for instance, route for its own connection map. But there are elements of routing and optimization that _could_ currently be performed by the quantum device itself  in **very limited** cases, and at a lower level of efficiency than performing this work on a classical computer.

The experiment of using the quantum device as its own optimizer seems worthwhile; it seems inevitable that as quantum hardware and computing science evolves, this routing and optimization will come to be performed by the quantum device itself.

I inquired of Dr. Bello whether there were distinct, identifiable components of the current transpiler technology amenable to experimentation of this sort.

Dr. Bello suggested [bit_mapping.py](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/passes/routing/bip_mapping.py) and [csp_layout.py](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/passes/layout/csp_layout.py)

**Reply by 1ucian0** [reply]
Another one, VF2Layout uses subgraph isomorphism: https://medium.com/qiskit/introducing-a-quantum-subgraph-isomorphism-algorithm-7bd860415422
I didn't give it a deep thought. Just dropping it here.

**Comment by jwoehr** [comment]
Possibly next mentorship program Spring 2022, I'll propose some work on the Transpiler.

**Comment by jwoehr** [comment]
See the QAMP Spring 2022 mentorship [Transpiler Hackathon](https://github.com/qiskit-advocate/qamp-spring-22/issues/1)
